rmf_fleet_adapter/src/rmf_fleet_adapter/events/internal_ChopeNegotiator.hpp

```diff
@@ -0,0 +1,350 @@
+          "Claim issued");
+      });
+
+
+    negotiator->_reservation_allocation =
+      negotiator->_context->node()->allocated_claims_obs().observe_on(rxcpp::identity_same_worker(
+          negotiator->_context->worker()))
+      .subscribe([w =
+        negotiator->weak_from_this()](const std::shared_ptr<rmf_chope_msgs::msg::ReservationAllocation>
+        &msg)
+        {
+          const auto self = w.lock();
+          if (!self)
+            return;
+
+          if (!self->_ticket.has_value())
+          {
+            return;
+          }
+
+          if (msg->ticket.ticket_id != self->_ticket.value()->ticket_id)
+          {
+            return;
+          }
+
+          self->_final_allocated_destination = msg;
+          self->_context->_set_allocated_destination(*msg.get());
+
+          if (self->_current_reservation_state ==  ReservationState::Requested)
+          {
+            while (auto allocation = self->_context->_release_resource())
+            {
+              rmf_chope_msgs::msg::ReleaseRequest msg;
+              std::stringstream str;
+              str << self->_context->location()[0].waypoint();
+              str >> msg.location;
+              msg.ticket = allocation->ticket;
+              self->_context->node()->release_location()->publish(msg);
+              RCLCPP_ERROR(
+                self->_context->node()->get_logger(),
+                "Releasing waypoint"
+              );
+            }
+          }
+
+          if (msg->instruction_type
+            == rmf_chope_msgs::msg::ReservationAllocation::IMMEDIATELY_PROCEED)
+          {
+            RCLCPP_INFO(
+                self->_context->node()->get_logger(), "chope: Robot %s is going to final destination",
+                  self->_context->name().c_str());
+            self->_current_reservation_state = ReservationState::RecievedResponseProceedImmediate;
+            self->_selected_final_destination_cb(self->_goals[self->_final_allocated_destination.value()->
+              satisfies_alternative]);
+          }
+
+          if (msg->instruction_type
+            == rmf_chope_msgs::msg::ReservationAllocation::WAIT_PERMANENTLY)
+          {
+            self->_current_reservation_state = ReservationState::RecievedResponceProceedWaitPoint;
+            self->_selected_waitpoint_cb(self->_waitpoints[self->_final_allocated_destination.value()->
+              satisfies_alternative]);
+            RCLCPP_INFO(
+                self->_context->node()->get_logger(), "chope: Robot %s is being asked to proceed to a to waitpoint",
+                  self->_context->name().c_str());
+          }
+        });
+
+    RCLCPP_INFO(context->node()->get_logger(),
+      "Sending chope request");
+    negotiator->make_request(same_map);
+    return negotiator;
+  }
+
+ChopeNodeNegotiator()
+  {
+
+  }
+private:
+  enum class ReservationState
+  {
+    Pending=0,
+    Requested=1,
+    RecievedResponceProceedWaitPoint=2,
+    RecievedResponseProceedImmediate=3
```

Comment:
Typo (I'd suggest just doing a search and replace in the whole diff for `Recieved -> Received`).